doc/developer/design/20230210_storage_persist_sink_redux.md

```diff
@@ -0,0 +1,195 @@
+                             | one arbitrary worker |
+                             +------+---------------+
+```
+
+The different operators work like this:
+
+1. `mint_batch_descriptions` emits new batch descriptions whenever the frontier of source `Collection` advances.
+A batch description is a pair of `(previous_upper, upper)` that tells write operators which updates to write and
+in the end tells the append operator what frontiers to use for the `compare_and_append`. This is a single-worker operator.
+2. `write_batches` writes the source `Collection` to persist. It does this by creating a persist `BatchBuilder` for each
+timestamp, and writing every value that occurs at each timestamp to the corresponding `BatchBuilder`. This uses a bounded
+amount of memory. This does not yet append the batches to the persist shard, the update are only uploaded/prepared to be
+appended to a shard. We only emit finalized batches to the `append_batches` operator when that batch is contained within
+a batch descriptions that we learned about from `mint_batch_descriptions`.
+3. `append_batches` takes as input the minted batch descriptions and written batches. Whenever the frontiers sufficiently advance,
+we take a batch description and all the batches that belong to it and append it to the persist shard.
+This also updates the shared frontier information in `storage_state` to match the persist frontier.
+This way, the controller eventually learns of the upper frontier of data written to the persist shard.
+Note that failing to `compare_and_append` must be retried, as we current expect only 1 replica to be
+running this sink operation for each source `Collection`.
+
+### Notes on the implementation
+
+- It is needlessly expensive to create a `BatchBuilder` for each timestamp.
+  - A change to this implementation is to avoid spilling data to persist for timestamps
+  whose values use a small amount of memory. These can be consolidated into large batches in `append_batches`,
+  similar to how <https://github.com/MaterializeInc/materialize/pull/17526> works.
+  - We do this in the current implementation, so doing it from many workers, instead of 1,
+  which has strictly more throughput (though does cause additional `SET` and `GET` load on AWS S3).
+  - Its undecided whether or not this should be required for the reference implementation.
+- `append_batches` and `write_batches` will be required to track statistics to expose to users.
+  - See <https://www.notion.so/materialize/User-facing-metrics-for-Sources-9603b61224a04a749969702e04662ed3> for more details
+
+
+# Rollout
+[rollout]: #rollout
+
+## Testing and observability
+[testing-and-observability]: #testing-and-observability
+
+This new implementation can be tested entirely using the existing CI, nightly, and release-qualification pipelines.
+However its important that the performance of the new implementation is carefully monitored as part of the first
+release that contains it.
```

Comment:
This sounds scary! What's the escape hatch should we determine the new implementation doesn't work in production?

Reply:
added some color in and update!